tools/testing/test_selections.py

```diff
@@ -67,97 +67,182 @@ def get_with_pytest_shard(
+    del tests
+
     known_tests = get_with_pytest_shard(known_tests, test_file_times, test_class_times)
 
     if sort_by_time:
         known_tests = sorted(known_tests, key=lambda j: j.get_time(), reverse=True)
 
-    sharded_jobs: List[ShardJob] = [ShardJob() for _ in range(num_shards)]
-    for test in known_tests:
-        if must_serial(test.name):
-            min_sharded_job = min(sharded_jobs, key=lambda j: j.get_total_time())
+    def _shard_serial(tests: List[ShardedTest], sharded_jobs: List[ShardJob]) -> None:
+        assert estimated_time_limit is not None, "Estimated time limit must be provided"
+        new_sharded_jobs = sharded_jobs
+        for test in tests:
+            if (
+                len(sharded_jobs) > 1
+                and sharded_jobs[-1].get_total_time() > estimated_time_limit
+            ):
+                new_sharded_jobs = sharded_jobs[:-1]
+            min_sharded_job = min(new_sharded_jobs, key=lambda j: j.get_total_time())
             min_sharded_job.serial.append(test)
-        else:
+
+    def _shard_parallel(tests: List[ShardedTest], sharded_jobs: List[ShardJob]) -> None:
+        for test in tests:
             min_sharded_job = min(sharded_jobs, key=lambda j: j.get_total_time())
             min_sharded_job.parallel.append(test)
 
+    if serial:
+        _shard_serial(known_tests, sharded_jobs)
+    else:
+        _shard_parallel(known_tests, sharded_jobs)
+
     # Round robin the unknown jobs starting with the smallest shard
+    num_shards = len(sharded_jobs)
     index = min(range(num_shards), key=lambda i: sharded_jobs[i].get_total_time())
     for unknown_test in unknown_tests:
         sharded_jobs[index].serial.append(ShardedTest(unknown_test, 1, 1, None))
         index = (index + 1) % num_shards
+
+    return
+
+
+def calculate_shards(
+    num_shards: int,
+    tests: Sequence[TestRun],
+    test_file_times: Dict[str, float],
+    test_class_times: Optional[Dict[str, Dict[str, float]]],
+    must_serial: Optional[Callable[[str], bool]] = None,
+    sort_by_time: bool = True,
+) -> List[Tuple[float, List[ShardedTest]]]:
+    must_serial = must_serial or (lambda x: True)
+    test_class_times = test_class_times or {}
+    serial_tests = [
+        test
+        for test in tests
+        if get_duration(test, test_file_times, test_class_times) is None
+        or must_serial(test.test_file)
+    ]
+    parallel_tests = [test for test in tests if test not in serial_tests]
+
+    serial_time = sum(
+        get_duration(test, test_file_times, test_class_times) or 0
+        for test in serial_tests
+    )
+    parallel_time = sum(
+        get_duration(test, test_file_times, test_class_times) or 0
+        for test in parallel_tests
+    )
+    total_time = serial_time + parallel_time / NUM_PROCS_FOR_SHARDING_CALC
+    estimated_time_per_shard = total_time / num_shards
+    estimated_time_limit = 0.0
+    if estimated_time_per_shard != 0:
```

Comment:
Appreciate if some comment could be added here to understand how the estimated time limit is calculated